// @elastic/apm-rum-core/dist/lib/index.js
"use strict";

exports.__esModule = true;
exports.createServiceFactory = createServiceFactory;

var _errorLogging = _interopRequireDefault(require("./error-logging"));

var _performanceMonitoring = _interopRequireDefault(require("./performance-monitoring"));

var _serviceFactory = _interopRequireDefault(require("./common/service-factory"));

exports.ServiceFactory = _serviceFactory.default;

var _utils = require("./common/utils");

exports.isPlatformSupported = _utils.isPlatformSupported;
exports.scheduleMicroTask = _utils.scheduleMicroTask;

var _patching = require("./common/patching");

exports.patchAll = _patching.patchAll;
exports.patchEventHandler = _patching.patchEventHandler;

var _constants = require("./common/constants");

exports.PAGE_LOAD = _constants.PAGE_LOAD;
exports.ERROR = _constants.ERROR;

var _instrument = require("./common/instrument");

exports.getInstrumentationFlags = _instrument.getInstrumentationFlags;

var _opentracing = require("./opentracing");

exports.createTracer = _opentracing.createTracer;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createServiceFactory() {
  var serviceFactory = new _serviceFactory.default();
  serviceFactory.registerCoreServices();

  _errorLogging.default.registerServices(serviceFactory);

  _performanceMonitoring.default.registerServices(serviceFactory);

  return serviceFactory;
}
