// @elastic/apm-rum-core/dist/lib/common/constants.js
"use strict";

exports.__esModule = true;
exports.TRANSACTION_TYPE_ORDER = exports.TEMPORARY_TYPE = exports.SIMILAR_SPAN_TO_TRANSACTION_RATIO = exports.BROWSER_RESPONSIVENESS_BUFFER = exports.BROWSER_RESPONSIVENESS_INTERVAL = exports.SERVER_URL_PREFIX = exports.KEYWORD_LIMIT = exports.HTTP_REQUEST_TYPE = exports.LOCAL_CONFIG_KEY = exports.AFTER_EVENT = exports.BEFORE_EVENT = exports.ERROR = exports.HISTORY = exports.FETCH = exports.XMLHTTPREQUEST = exports.CONFIG_CHANGE = exports.TRANSACTION_END = exports.TRANSACTION_START = exports.USER_TIMING_THRESHOLD = exports.TYPE_CUSTOM = exports.NAME_UNKNOWN = exports.ROUTE_CHANGE = exports.PAGE_LOAD = exports.MAX_SPAN_DURATION = exports.REUSABILITY_THRESHOLD = exports.RESOURCE_INITIATOR_TYPES = exports.REMOVE_EVENT_LISTENER_STR = exports.ADD_EVENT_LISTENER_STR = exports.CLEAR = exports.INVOKE = exports.SCHEDULE = void 0;
var SCHEDULE = 'schedule';
exports.SCHEDULE = SCHEDULE;
var INVOKE = 'invoke';
exports.INVOKE = INVOKE;
var CLEAR = 'clear';
exports.CLEAR = CLEAR;
var ADD_EVENT_LISTENER_STR = 'addEventListener';
exports.ADD_EVENT_LISTENER_STR = ADD_EVENT_LISTENER_STR;
var REMOVE_EVENT_LISTENER_STR = 'removeEventListener';
exports.REMOVE_EVENT_LISTENER_STR = REMOVE_EVENT_LISTENER_STR;
var RESOURCE_INITIATOR_TYPES = ['link', 'css', 'script', 'img', 'xmlhttprequest', 'fetch', 'beacon', 'iframe'];
exports.RESOURCE_INITIATOR_TYPES = RESOURCE_INITIATOR_TYPES;
var REUSABILITY_THRESHOLD = 5000;
exports.REUSABILITY_THRESHOLD = REUSABILITY_THRESHOLD;
var MAX_SPAN_DURATION = 5 * 60 * 1000;
exports.MAX_SPAN_DURATION = MAX_SPAN_DURATION;
var PAGE_LOAD = 'page-load';
exports.PAGE_LOAD = PAGE_LOAD;
var ROUTE_CHANGE = 'route-change';
exports.ROUTE_CHANGE = ROUTE_CHANGE;
var TYPE_CUSTOM = 'custom';
exports.TYPE_CUSTOM = TYPE_CUSTOM;
var HTTP_REQUEST_TYPE = 'http-request';
exports.HTTP_REQUEST_TYPE = HTTP_REQUEST_TYPE;
var TEMPORARY_TYPE = 'temporary';
exports.TEMPORARY_TYPE = TEMPORARY_TYPE;
var NAME_UNKNOWN = 'Unknown';
exports.NAME_UNKNOWN = NAME_UNKNOWN;
var TRANSACTION_TYPE_ORDER = [PAGE_LOAD, ROUTE_CHANGE, HTTP_REQUEST_TYPE, TYPE_CUSTOM, TEMPORARY_TYPE];
exports.TRANSACTION_TYPE_ORDER = TRANSACTION_TYPE_ORDER;
var USER_TIMING_THRESHOLD = 60;
exports.USER_TIMING_THRESHOLD = USER_TIMING_THRESHOLD;
var TRANSACTION_START = 'transaction:start';
exports.TRANSACTION_START = TRANSACTION_START;
var TRANSACTION_END = 'transaction:end';
exports.TRANSACTION_END = TRANSACTION_END;
var CONFIG_CHANGE = 'config:change';
exports.CONFIG_CHANGE = CONFIG_CHANGE;
var XMLHTTPREQUEST = 'xmlhttprequest';
exports.XMLHTTPREQUEST = XMLHTTPREQUEST;
var FETCH = 'fetch';
exports.FETCH = FETCH;
var HISTORY = 'history';
exports.HISTORY = HISTORY;
var ERROR = 'error';
exports.ERROR = ERROR;
var BEFORE_EVENT = ':before';
exports.BEFORE_EVENT = BEFORE_EVENT;
var AFTER_EVENT = ':after';
exports.AFTER_EVENT = AFTER_EVENT;
var LOCAL_CONFIG_KEY = 'elastic_apm_config';
exports.LOCAL_CONFIG_KEY = LOCAL_CONFIG_KEY;
var KEYWORD_LIMIT = 1024;
exports.KEYWORD_LIMIT = KEYWORD_LIMIT;
var SERVER_URL_PREFIX = '/intake/v2/rum/events';
exports.SERVER_URL_PREFIX = SERVER_URL_PREFIX;
var BROWSER_RESPONSIVENESS_INTERVAL = 500;
exports.BROWSER_RESPONSIVENESS_INTERVAL = BROWSER_RESPONSIVENESS_INTERVAL;
var BROWSER_RESPONSIVENESS_BUFFER = 3;
exports.BROWSER_RESPONSIVENESS_BUFFER = BROWSER_RESPONSIVENESS_BUFFER;
var SIMILAR_SPAN_TO_TRANSACTION_RATIO = 0.05;
exports.SIMILAR_SPAN_TO_TRANSACTION_RATIO = SIMILAR_SPAN_TO_TRANSACTION_RATIO;
